module/zfs/zfs_vfsops.c

```diff
@@ -1944,6 +1948,7 @@ zfs_umount(struct super_block *sb)
 		dmu_objset_disown(os, B_TRUE, zfsvfs);
 	}
 
+	dataset_kstats_destroy(&zfsvfs->z_kstat);
```

Comment:
This will be called twice, once here and a second time in `zfsvfs_free()`.   What do you think about moving the `dataset_kstats_create()` call in to the `if (mounting) { }` conditional of `zfsvfs_setup()`.  This way you'll only need the single call in `zfsvfs_free()` to cover all the cleanup paths.